### PR TITLE
Bennie/mraa pinmux fix

### PR DESCRIPTION
**Pull request by benbrenson**
This pull requests covers roughly the following topics:

* Add patches to mraa for supporting pinmuxing via debugfs rather than mmap'ed access to pad configuration registers
* Add patches (linux-iot2050) for supporting the userspace pinmux feature (available since v5.13)
* Add one patch (linux-iot2050) to adapt the iot20250 device tree to support the userspace pinmux feature 

I see that there are a lot of changes to persistent kernel patch series which have changed due to my workflow. Changes only apply to nothing but the commit hashes. During kernel development I fetched the cip kernel and applied all persistent patches and then recreated them with my additional patchset. 
Please don't hesitate to correct me here!

My patchset starts at "0211-pinctrl-pinmux-add-function-selector-to-pinmux-funct.patch".

### CODE REVIEW

**Comment by jan-kiszka** [comment]
Please generate the kernel patch queue with `--no-numbered --no-signature`.

**Comment by benbrenson** [comment]
Ok thanks. 
That reduced the diff a lot. But there are still some problems.
This patch [here](https://github.com/siemens/meta-iot2050/pull/437/files#diff-6637f884470a17dac480382c7c6453bc78bc90a0abe0e3a004b43a512000d8d1) (0048-*) is completely updated. 
Seems like some changes of this patch made it already into upstream?
Or I took the wrong baseline?
Or my workflow breaks things here?

What I did:

```bash
git clone https://gitlab.com/cip-project/cip-kernel/linux-cip.git
cd linux-cip && git checkout v5.10.162-cip24
```
Apply all patches found in patches-5.10/:
```bash
find patches-5.10/* | while read file; do cat "$file" | git am ; done
```
All patches applied without any issues. 

After that I did my work and regenerated the new patchseries afterwards:
```bash
git format-patch --no-numbered --no-signature v5.10.162-cip24..HEAD -o patches/to_isar/
```

**Comment by benbrenson** [comment]
Ok I found the issue.

If I append "-C" to the git format-patch command it works. Seems like the copy-detection was turned off on my system.

File: recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch [hunk 1]
```diff
@@ -0,0 +1,754 @@
+From be89e6880a715a38405cec22364754661834aa11 Mon Sep 17 00:00:00 2001
+From: Benedikt Niedermayr <benedikt.niedermayr@siemens.com>
+Date: Tue, 21 Feb 2023 11:03:42 +0100
+Subject: [PATCH] arm64: dts: ti: Definitions for runtime pinmuxing
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch:4]: You should address the iot2050 with the subject.

File: recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+ 		pinctrl-single,pins = <
+ 			AM65X_IOPAD(0x0000, PIN_OUTPUT, 1)  /* VOUT1_DATA0 */
+@@ -477,14 +1076,7 @@ &main_uart2 {
+ 	status = "disabled";
+ };
+ 
+-&mcu_uart0 {
+-	pinctrl-names = "default";
+-	pinctrl-0 = <&arduino_uart_pins_default>;
+-};
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch:716]: Is this change related?

**Comment by jan-kiszka** [comment]
Ready for testing, I would say. Once we are fine, changes should also be proposed upstream.

For the 5.10 backports, we could think about proposing them for the CIP kernel. OTOH, they are all part of 6.1, right? That will be our next kernel anyway.

File: recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch [hunk 3]
```diff
@@ -0,0 +1,754 @@
+From be89e6880a715a38405cec22364754661834aa11 Mon Sep 17 00:00:00 2001
+From: Benedikt Niedermayr <benedikt.niedermayr@siemens.com>
+Date: Tue, 21 Feb 2023 11:03:42 +0100
+Subject: [PATCH] arm64: dts: ti: Definitions for runtime pinmuxing
+
+Multiple device tree nodes where added in order to support
+runtime pinmuxing via debugfs.
```
[the remark this answers is not in the document]
Comment (benbrenson) [reply recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch:7]: Ok. Makes sense

File: recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch (hunk 1 above)
Comment (benbrenson) [reply recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch:4]: So I change that to  "iot2050: Definitions for runtime pinmuxing" ?

File: recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch (hunk 2 above)
Comment (benbrenson) [reply recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch:716]: Yes, this was intended.
Maybe a short explanation:

I had to add all possible pinmux states for each pin into the related pinctrl node. Otherwise these states are not visible for the pinctrl node. 

This affected pins that have already been configured and assigned to interfaces like I2C, SPI and UART. Those pins and their states had to be moved to the pinctrl node. 
Otherwise we will have ownership issues. 
The kernel would complain about pins that have already been assigned to the UART device in this specific case.

File: recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch (hunk 1 above)
Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-5.10/0220-arm64-dts-ti-Definitions-for-runtime-pinmuxing.patch:4]: arm64: dts: ti: iot2050: ...

**Comment by benbrenson** [comment]
> Ready for testing, I would say. Once we are fine, changes should also be proposed upstream.
> 
> For the 5.10 backports, we could think about proposing them for the CIP kernel. OTOH, they are all part of 6.1, right? That will be our next kernel anyway.

Yes. The pinctrl backports have been picked from v5.13. So they can be removed for 6.1.

File: recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
++iot2050_mux_debugfs(const char *base_dir, const char *group, const char *function, const char* gpio_mode)
++{
++    FILE *fd = NULL;
++    char p_pinmux[PATH_MAX];
++    char mux[MRAA_PIN_NAME_SIZE];
++    int ret;
++
++    syslog(LOG_DEBUG, "iot2050: debugfs: enter\n");
++
++    if(!base_dir || !group || !function)
```
Comment (jan-kiszka) [inline recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:57]: `if (...)` - `if` is not a function

File: recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
++        return MRAA_SUCCESS;
```
Comment (jan-kiszka) [inline recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:58]: Success or rather error?

File: recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+
+This patch adds support for multiplexing pins via debugfs rather
+than access memory mapped pad-configuration registers.
+The debugfs pinmux offers the possbility to run mraa on iot2050
+platforms as regular user instead of root by adjusting privileges
+on debugfs files.
+
+Bias settings are currently also configured by accessing pinmux.
+Unfortunatelly a proper upstream-like pinconf usage is currently
+not possible.
```
Comment (jan-kiszka) [inline recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:14]: Am I right the you leave the /dev/mem bases muxing support in place? I don't see related code being removed. Is that as fallback on systems not yet support the new scheme?

Comment (benbrenson) [reply recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:14]: Yes, that was one idea behind that.
The other idea was to remove the /dev/mem access in a separate commit. Depends on discussions and opinions upstream with MRAA folks... 

The fallback mechanism would be a valid/interesting use case if someone uses MRAA as standalone on custom kernels (not directly provided by meta-iot2050).

But in case someone uses the image produced by meta-iot2050 this fallback implementation makes no sense, since all required DTB changes are already included.

File: recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch (hunk 5 above)
Comment (benbrenson) [reply recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:58]: Error. I'll fix that.

File: recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch (hunk 4 above)
Comment (benbrenson) [reply recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:57]: Right. But MRAA seems to not follow that convention. 
Nevertheless, I see that I didn't use that consistently throughout iot2050_mux_debugfs().

File: recipes-kernel/linux/files/patches-5.10/0001-dmaengine-ti-k3-udma-glue-Add-function-to-get-device.patch [hunk 7]
```diff
@@ -1,4 +1,4 @@
-From 3f592d3a8794d4db3192b71a7f8df363cb1c2c62 Mon Sep 17 00:00:00 2001
+From 274d819974b7cb1cfadf29a6312d0ede1119a8af Mon Sep 17 00:00:00 2001
```
Comment (BaochengSu) [inline recipes-kernel/linux/files/patches-5.10/0001-dmaengine-ti-k3-udma-glue-Add-function-to-get-device.patch:1]: Per my experience, this change could be safely ignored.

BTW, @jan-kiszka , I think it might be a good idea to apply the --zero-commit option, see https://github.com/git/git/blame/master/Documentation/git-format-patch.txt#LL384C16-L384C16. What do you think?

Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-5.10/0001-dmaengine-ti-k3-udma-glue-Add-function-to-get-device.patch:1]: Good suggestion, indeed.

Comment (benbrenson) [reply recipes-kernel/linux/files/patches-5.10/0001-dmaengine-ti-k3-udma-glue-Add-function-to-get-device.patch:1]: Yes, good idea.

File: recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch (hunk 6 above)
Comment (benbrenson) [reply recipes-app/mraa/files/0003-iot2050-add-debugfs-pinmux-and-gpio-chardev-support.patch:14]: But I'll go with that fallback mechanism for now

**Comment by jan-kiszka** [comment]
Looks good otherwise.

**Comment by benbrenson** [comment]
Sorry for the late reply. I missed your comment and I was sick last week.

I rebased to current master and realigned the patchseries. Meanwhile [BaochengSu](https://github.com/BaochengSu) already added the "--zero-commit" Option, so I used that as well.

**Comment by jan-kiszka** [comment]
Roadmap should likely be now:
1. merge
2. test extensively (system test)
3. propose changes upstream (kernel, mraa)
